### PR TITLE
fault: only check for overflow when performing faults

### LINKED ISSUE #12816: [FaultFilter] Inaccurate faults overflow metrics
### Intro

One of the first operations  Envoy's fault filter performs when it receives an incoming network request is [checking for whether the maximum number of active faults has been reached or not](https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/http/fault/fault_filter.cc#L121). If the number of currently active faults is greater than agreed on maximum limit, [Envoy increments `faults_overflow` counter](https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/http/fault/fault_filter.cc#L216) which is supposed to help its users to understand when too many flows are flowing into the system.

### Issue

Envoy increases the value of `faults_overflow` counter without checking whether an incoming requests "contains" a fault or not. This results in `faults_overflow` metric being inaccurate. Whether `faults_overflow` metrics is incremented or not for cases when fault filter is serving the maximum number of faults, should depend on whether Envoy was instructed to inject a fault into a given incoming network request - the one being processed at the moment. If there is no fault to inject, the `faults_overflow` metric should not be incremented even if the number of `active_faults` is at its maximum value.

### PR DESCRIPTION
**Pull request by snowp**
This shuffles the check for overflow around so that we only increment
the overflow value when attempting to apply a fault instead of for
each request. This ensures that if applying a low % fault the overflow
stat will be in line with that, instead of counting each request once
the overflow threshold has been hit.

Signed-off-by: Snow Pettersen <snowp@lyft.com>

Risk Level: Medium
Testing: New UT, existing tests
Docs Changes: n/a
Release Notes: n/a
Fixes #12816

### CODE REVIEW

File: source/extensions/filters/http/fault/fault_filter.cc [hunk 1]
```diff
@@ -391,15 +401,29 @@ FaultFilterStats FaultFilterConfig::generateStats(const std::string& prefix, Sta
...
   if (fault_active_) {
-    return;
+    return true;
+  }
+
+  // We only check for overflow when attempting to perform a fault. Note that this means that a
+  // single request might increment the counter more than once if it tries to apply multiple faults,
+  // though it is also possible for it to fail the first check then succeed on the second (should
+  // another thread decrement the fault).
+  // TODO(snowp): Is this behavior ideal? Should we track whether we've rejected faults for this
```
Comment (snowp) [inline source/extensions/filters/http/fault/fault_filter.cc:416]: this is the main open question in my mind, not sure how important we feel this is

**Comment by mattklein123** [comment]
@Augustyniak can you take a first pass?

**Comment by Augustyniak** [comment]
> @Augustyniak can you take a first pass?

Sure.

File: test/extensions/filters/http/fault/fault_filter_test.cc [hunk 2]
```diff
@@ -511,18 +508,37 @@ TEST_F(FaultFilterTest, FixedDelayZeroDuration) {
...
+              featureEnabled("fault.http.delay.fixed_delay_percent",
+                             Matcher<const envoy::type::v3::FractionalPercent&>(Percent(100))))
+      .WillOnce(Return(true));
+
+  // Return 1ms delay
+  EXPECT_CALL(runtime_.snapshot_, getInteger("fault.http.delay.fixed_duration_ms", 5000))
+      .WillOnce(Return(1));
 
-  EXPECT_CALL(runtime_.snapshot_, getInteger("fault.http.max_active_faults", 0))
+  EXPECT_CALL(runtime_.snapshot_,
```
Comment (Augustyniak) [inline test/extensions/filters/http/fault/fault_filter_test.cc:523]: in line #528 - can we verify that active_faults counter is equal to 0 when there is a faults_overflow.

File: source/extensions/filters/http/fault/fault_filter.cc [hunk 3]
```diff
@@ -391,15 +401,29 @@ FaultFilterStats FaultFilterConfig::generateStats(const std::string& prefix, Sta
                                  POOL_GAUGE_PREFIX(scope, final_prefix))};
 }
 
-void FaultFilter::maybeIncActiveFaults() {
+bool FaultFilter::maybeDoFault() {
```
[the remark this answers is not in the document]
Comment (snowp) [reply source/extensions/filters/http/fault/fault_filter.cc:404]: Yeah I tried to capture the fact that we also increment the gauge here, but I agree that it's a bit misleading. I'll go with your suggestion

[the remark this answers is not in the document]
Comment (snowp) [reply source/extensions/filters/http/fault/fault_filter.cc:404]: Actually will do `tryIncActiveFaults` which I think is a bit clearer

File: source/extensions/filters/http/fault/fault_filter.cc (hunk 1 above)
Comment (mattklein123) [reply source/extensions/filters/http/fault/fault_filter.cc:416]: IMO it's probably fine either way. I would go with whatever is easiest and most clear to implement. Can you remove the TODO and make it clear what we have chosen? Also not sure if it's worth it to add docs on the stat that makes this mroe clear, up to you.

Comment (snowp) [reply source/extensions/filters/http/fault/fault_filter.cc]: I think the docs imply the behavior I changed it to - it counts the number of *faults* were skipped due to overflow - so I'll leave it at that. Updated the comment a bit and removed the TODO

**Pull request merged**